src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -63,8 +63,36 @@ void GcsNodeManager::HandleRegisterNode(rpc::RegisterNodeRequest request,
     AddNode(std::make_shared<rpc::GcsNodeInfo>(request.node_info()));
     GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
   };
-  RAY_CHECK_OK(
-      gcs_table_storage_->NodeTable().Put(node_id, request.node_info(), on_done));
+  if (request.node_info().is_head_node()) {
+    // mark all old head nodes as dead if exists:
+    // 1. should never happen when HA is not used
+    // 2. happens when a new GCS is started
+
+    std::vector<NodeID> head_nodes;
+    for (auto &node : alive_nodes_) {
+      if (node.second->is_head_node()) {
+        head_nodes.push_back(node.first);
+      }
+    }
```

Comment:
After a second thought, I'd prefer to have a RAY_CHECK to make sure there is at most one alive head node. I think this is guaranteed now since we always mark the old one as dead before adding the new head.

I think this guarantee makes reasoning the code easier.

Thoughts @rkooo567 @jonathan-anyscale ?

Reply:
I can agree with this, so we can emphasize the reason we added this code

Reply:
yeah that sounds good to me